### ISSUE TITLE
buf:lint:ignore should sometimes cascade

### ISSUE THREAD

**Issue by powerman** [bufbuild/buf#242]
I've tried to use it several times, in different situations and at different levels - it never work for me. Maybe I'm doing something wrong? Here is last case and how to reproduce:

```
$ cd $(mktemp -d)
$ cat >buf.yaml <<EOF
version: v1beta1
lint:           
  except:
    - PACKAGE_VERSION_SUFFIX        # Allow omitted version for v1.
    - RPC_RESPONSE_STANDARD_NAME    # Allow CRUD to return entity type.
  service_suffix: Svc                                                  
  allow_comment_ignores: true
EOF
$ mkdir example
$ cat >example/service.proto <<EOF
syntax = "proto3";

package example;  

option go_package = "example";

// Service.                   
service ExampleSvc {
    // buf:lint:ignore RPC_REQUEST_STANDARD_NAME
    rpc ReverseRPC(stream ReverseRPCResponse) returns (stream ReverseRPCRequest);
}                                                                                

// Request.
message ReverseRPCRequest {}

// Response.                
message ReverseRPCResponse {}
EOF
$ buf lint
example/service.proto:10:27:RPC request type "ReverseRPCResponse" should be named "ReverseRPCRequest" or "ExampleSvcReverseRPCRequest".
```

The rationale why I'm trying to disable this check at method level is: this method does 2-way streaming to call gRPC methods on client instead of server, i.e. client connects to server and then read stream of requests, executes them, and sends to server stream of responses - so while in general I prefer to have RPC_REQUEST_STANDARD_NAME enabled in this concrete case it'll result in very bad names with opposite meaning, which is not acceptable.

**Comment by powerman** [comment]
This happens on buf since at least v0.33.0 (didn't tried older ones) up to current 0.35.1.

**Comment by bufdev** [comment]
Will investigate

**Comment by bufdev** [comment]
So in short, the issue is that the linter looks at the attached comments (as defined by the Protobuf compiler) to the element under violation. With your above example, the comment ignoring `RPC_REQUEST_STANDARD_NAME` is attached to the `rpc`, not to the definition of what the request type is. It gets worse if you define it as streaming, as if you put the comment above `stream Request`, then it attaches to `stream`, not `Request`.

This, however will work:

```proto
syntax = "proto3";

package example;

service ExampleSvc {
    rpc ReverseRPC(
      stream
      // buf:lint:ignore RPC_REQUEST_STANDARD_NAME
      ReverseRPCResponse
    ) returns (stream ReverseRPCRequest);
}

message ReverseRPCRequest {}

message ReverseRPCResponse {}
```

We can understand why this is confusing. We'll look into seeing if we can make this more intuitive, although not high priority.

**Comment by powerman** [comment]
Thanks. To me it looks like it should always cascade, not sometimes - this is usual and most intuitive behaviour for linters and often helps to avoid useless duplications like in #161.

**Comment by powerman** [comment]
On a side note, to me it looks unclear is it good idea to apply RPC_REQUEST_STANDARD_NAME and RPC_RESPONSE_STANDARD_NAME to streams at all - both because streams rarely send "request" and "response", they're usually sending "just a message", and because these "standard names" originally wasn't designed for a streams at all (in fact, I don't see any recommendations about how to use streams in https://google.aip.dev/ at all). If you agree I can open another issue to disable these options for streams by default.

**Comment by bufdev** [comment]
We don't think that streams represent a special case here, the same concerns exist for request/response types with streaming RPCs as for unary RPCs. For this and many other reasons, we don't think aip.dev is authoritative for best practices in modern Protobuf development.

We'll look into fixing the cascade issue.

**Comment by bufdev** [comment]
See https://github.com/bufbuild/buf/pull/243

**Comment by bufdev** [comment]
This is fixed on master, and will go out in the next release (probably on Monday). If you want to check it out beforehand, you can do `go get github.com/bufbuild/buf/cmd/buf@master`. Thanks for filing the issue!

**Comment by powerman** [comment]
Thanks!

> For this and many other reasons, we don't think aip.dev is authoritative for best practices in modern Protobuf development.

Is there any other resource which you can recommend instead of aip.dev?

**Comment by bufdev** [comment]
Yes :-)

- https://docs.buf.build/lint-rules
- https://docs.buf.build/style-guide

This is released in v0.36.0.

**Comment by powerman** [comment]
Well, to be honest, your style guide covers only very small subset of things covered by AIP (and, at a glance, match AIP recommendations). I can't say I'm happy with everything AIP says, but I believe it's better to follow common style even if don't like it than invent your own. If someone will provide more suitable style guide (taking in account "your are not Google" idea) with amount of covered topics similar to AIP - I'll be happy to use it, but for now following AIP sounds more reasonable choice.
